internal/userutils/userutils.go

```diff
@@ -3,38 +3,71 @@ package userutils
 
 import (
 	"errors"
-	"fmt"
-	"os"
+	"sync/atomic"
+
+	"github.com/ubuntu/authd/internal/testsdetection"
 )
 
-// GroupFile is the path to the group file. It is exported for testing purposes.
-var GroupFile = "/etc/group"
-
-// LockGroupFile creates a lock file at /etc/group.lock, the same location
-// used by tools like gpasswd and groupmod to prevent concurrent modifications
-// to the /etc/group file.
-// The lock file contains the PID of the process that created it.
-func LockGroupFile() (err error) {
-	lockFile, err := os.OpenFile(GroupFile+".lock", os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0600)
-	if err != nil {
-		return fmt.Errorf("failed to open lock file: %w", err)
-	}
-	defer func() {
-		closeErr := lockFile.Close()
-		err = errors.Join(err, closeErr)
-	}()
+var (
+	writeLockShadowPasswordImpl   = writeLockShadowPassword
+	writeUnlockShadowPasswordImpl = writeUnlockShadowPassword
 
-	if _, err := lockFile.WriteString(fmt.Sprintf("%d", os.Getpid())); err != nil {
-		return fmt.Errorf("failed to write PID to lock file: %w", err)
-	}
+	overrideLocked atomic.Bool
+)
 
-	return nil
+var (
+	// ErrLock is the error when locking the database fails.
+	ErrLock = errors.New("failed to lock the shadow password database")
+
+	// ErrUnlock is the error when unlocking the database fails.
+	ErrUnlock = errors.New("failed to unlock the shadow password database")
+)
+
+// WriteLockShadowPassword locks for writing the shadow password database by
+// using the standard libc lckpwdf() function.
+// While the database is locked read operations can happen, but no other process
+// is allowed to write.
+// Note that this call will block all the other processes trying to access the
+// database in write mode, while it will return an error if called while the
+// lock is already hold by this process.
+func WriteLockShadowPassword() error {
```

Comment:
While the man page of `lckpwdf` states that it is for protecting the shadow password database, it's clearly used for more than that, i.e. protecting `/etc/passwd`, `/etc/shadow`, `/etc/group` and `/etc/gshadow`. So IMO we should not pretend that it is only for locking the shadow password database, and also remove "Shadow" from the function names.

The corresponding systemd function is named `take_etc_passwd_lock`. That makes sense as it seems that systemd only ever uses it to protect `/etc/passwd`. We currently only use it to protect `/etc/group`, so I think it would be sensible to name it `LockGroupFile`.

Reply:
We are currently using it for locking group file, but we will need to just lock everything while acting on users and groups, so I'd rather use a "generic name" now and avoid the refactor very soon; and indeed as you guessed I was using that name since it's what the manual refers.

Now, something like `WriteLockLocalAccountsDatabases`? It's very verbose... But...

Or maybe.... Just move all this into the namespace we already have for this stuff (`localentries`, in `internal/users`)  and just use `localentries.WriteLock`, `localentries.WriteUnlock`, as it should be quite self-explainatory

Reply:
> avoid the refactor very soon

why do you expect that we need to write to `/etc/passwd` or `/etc/shadow` very soon? I don't see why we would have to do that. We currently don't touch these at all.

Reply:
If I want to use this to test the behavior for local users, it will help...

Now we just use root for this, but root is still a special case that has different pam handling normally.

So I'd rather keep this flexible enough so that we can mock users better with it.

Reply:
Still not sure I understand, but it's probably not worth explaining more if you can just show me those tests soon :)

How about `LockPasswdAndGroupFiles`?

Reply:
Well, every time we're processing an user request we'll have to lock the local entries, to ensure that nothing else will try to interfere meanwhile, so this about locking everything.

I'd rather move everything to `localentries` and see what you think. Since that namespace is meant to handle local groups and users.

Reply:
> every time we're processing an user request we'll have to lock the local entries, to ensure that nothing else will try to interfere meanwhile

Wait what? Why do we have to lock anything on a user request? We only need `lckpwdf` because we're writing to `/etc/groups` directly when migrating to lowercase usernames, which is a one-time operation done during authd startup.

Reply:
I'm also getting the feeling we're spending way too much effort on making this migration absolutely safe, given that it only happens once for each existing authd installation. The chance that something else is writing to `/etc/groups` at the same time seems negligible.

Reply:
Let's close this conversation, for the public record the related resources were https://github.com/systemd/systemd/blob/main/docs/UIDS-GIDS.md?plain=1#L224-L233

On the naming side, let's move to https://github.com/ubuntu/authd/pull/723#discussion_r2063310609